store/tikv/backoff.go

```diff
@@ -128,15 +129,24 @@ type Backoffer struct {
 	maxSleep   int
 	totalSleep int
 	errors     []error
+	ctx        context.Context
 }
 
 // NewBackoffer creates a Backoffer with maximum sleep time(in ms).
-func NewBackoffer(maxSleep int) *Backoffer {
+func NewBackoffer(maxSleep int, ctx context.Context) *Backoffer {
 	return &Backoffer{
 		maxSleep: maxSleep,
+		ctx:      ctx,
 	}
 }
 
+// WithCancel returns a cancel function which, when called, would cancel backoffer's context.
+func (b *Backoffer) WithCancel() context.CancelFunc {
+	var cancel context.CancelFunc
+	b.ctx, cancel = context.WithCancel(b.ctx)
```

Comment:
standard use context.Context follow this pattern:

```
func f(ctx context.Context, ...) {
    ctx1, cancel := context.WithCancel(ctx)
    g(ctx1, ...)
}
```
our functions all use `bo *BackOffer` as the first argument, so I put ctx in BackOffer struct.

WithCancel won't overwrite parent context, it just wrap the parent context...

```
func f(bo *Backoffer, ...) {
  cancel := bo.WithCancel(ctx)
  g(bo, ...)
}
```

this code works just the same.